packages/core/src/external/carequality/pd/validating-pd.ts

```diff
@@ -0,0 +1,175 @@
+  constructor(message?: string) {
+    super(message);
+    this.name = "InternalError";
+  }
+}
+
+export function validateFHIRAndExtractPatient(patient: FHIRPatient): PatientData {
+  const firstName = patient.name?.[0]?.given?.[0]; // TODO we are taking the first index here but there might be multiple given names
+  if (!firstName) {
+    throw new InternalError("Given name is not defined");
+  }
+  const lastName = patient.name?.[0]?.family;
+  if (!lastName) {
+    throw new InternalError("Family name is not defined");
+  }
+  const birthDate = patient.birthDate;
+  if (!birthDate) {
+    throw new InternalError("Birth date is not defined");
+  }
+
+  const genderAtBirth =
+    patient.gender === "male" ? "M" : patient.gender === "female" ? "F" : undefined;
+  if (!genderAtBirth) {
+    throw new LivingSubjectAdministrativeGenderRequestedError("Gender at Birth is not defined");
+  }
+
+  const addresses = (patient.address ?? []).map((addr: FHIRAddress) => {
+    const addressLine1 = addr.line ? addr.line.join(" ") : "";
+    const city = addr.city || "";
+    const state = addr.state ? getStateEnum(addr.state) : undefined;
+    const zip = addr.postalCode || "";
+    const country = addr.country || "USA";
+
+    if (!addressLine1) {
+      throw new PatientAddressRequestedError("Address Line 1 is not defined");
+    }
+    if (!city) {
+      throw new PatientAddressRequestedError("City is not defined");
+    }
+    if (!state) {
+      throw new PatientAddressRequestedError("State is not defined");
+    }
+    if (!zip) {
+      throw new PatientAddressRequestedError("Zip is not defined");
+    }
+
+    const newAddress: Address = {
+      addressLine1,
+      city,
+      state,
+      zip,
+      country,
+    };
+    return newAddress;
+  });
+
+  const contacts = (patient.telecom ?? []).map((tel: ContactPoint) => {
+    const contact: Contact = {};
+    if (tel.system && isContactType(tel.system)) {
+      contact[tel.system] = tel.value;
+    }
+    return contact;
+  });
+
+  const personalIdentifiers = (patient.identifier ?? [])
+    .map((identifier: Identifier) => {
+      const system = identifier.system;
+      const value = identifier.value;
+      if (system && value && stateMappings[system]) {
+        const state = stateMappings[system];
+        if (state) {
+          const personalIdentifier: PersonalIdentifier = {
+            type: "driversLicense",
+            value: value,
+            state: state,
+          };
+          return personalIdentifier;
+        }
+      }
+      return undefined;
+    })
+    .filter((item): item is PersonalIdentifier => item !== undefined);
+
+  const convertedPatient: PatientData = {
+    firstName: firstName,
+    lastName: lastName,
+    dob: birthDate,
+    genderAtBirth: genderAtBirth,
+    address: addresses,
+    contact: contacts,
+    personalIdentifiers: personalIdentifiers,
+  };
+  return convertedPatient;
+}
+
+const stateMappings: { [key: string]: USState } = {
```

Comment:
This seem like worth it's own file.

Also, the way we're defining the type we're not validating at compile time that all states have an entry. If we build it the other way around, it would (`{ [key in USState]: string }`

Reply:
In downstream PR I moved it into a shared variables file for carequality.
The only issue with defining it the other way around is that we get the system (the string) from mirth but not the state, so mapping needs to go the way I originally defined it

Reply:
> The only issue with defining it the other way around is that we get the system (the string) from mirth but not the state, so mapping needs to go the way I originally defined it

We can just search by values instead of keys.

Reply:
In the downstream PR its in a shared file for all CQ things, not just PD things. 
I don't love that as a solution - seems janky and like contrary to the point of having a dictionary. As a compromise for now, I added a test. Will add it as a todo to downstream PR to come to final decision on:

```
describe('State Mappings Validation', () => {
  it('should have an entry for every USState', () => {
    const allStates = Object.values(USState);
    const mappedStates = new Set(Object.values(stateMappings));

    allStates.forEach(state => {
      expect(mappedStates.has(state)).toBeTruthy();
    });
  });
});
```